src/dmd/root/file.d

```diff
@@ -261,4 +261,116 @@ nothrow:
             assert(0);
         }
     }
+
+    /***************************************************
+     * Update file
+     *
+     * If the file exists and is identical to what is to be written,
+     * merely update the timestamp on the file.
+     * Otherwise, write the file.
+     *
+     * The idea is writes are much slower than reads, and build systems
+     * often wind up generating identical files.
+     * Params:
+     *  name = name of file to update
+     *  data = updated contents of file
+     * Returns:
+     *  `true` on success
+     */
+    extern (D) static bool update(const(char)* namez, const void[] data)
+    {
+        enum log = false;
+        if (log) printf("update %s\n", namez);
+        version (Windows)
+        {
+            const nameStr = namez.toDString();
+
+            import core.sys.windows.windows;
+
+            WIN32_FILE_ATTRIBUTE_DATA fad = void;
+            // Doesn't exist, not a regular file, different size
+            if (nameStr.extendedPathThen!(p => GetFileAttributesExW(p.ptr, GET_FILEEX_INFO_LEVELS.GetFileExInfoStandard, &fad)) == 0 ||
+                fad.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY ||
+                ((cast(ulong) fad.nFileSizeHigh << 32) | fad.nFileSizeLow) != data.length)
+            {
+                return write(namez, data);               // write new file
+            }
+        }
+        else version (Posix)
+        {
+            import core.sys.posix.sys.stat;
+
+            stat_t statbuf = void;
+            if (stat(namez, &statbuf) != 0 ||            // doesn't exist
+                (statbuf.st_mode & S_IFMT) != S_IFREG || // not a regular file
+                statbuf.st_size != data.length)          // different size
+            {
+                if (log) printf("not exist or diff size %d %d %d\n",
+                    stat(namez, &statbuf) != 0,
+                    (statbuf.st_mode & S_IFMT) != S_IFREG,
+                     statbuf.st_size != data.length);
+                return write(namez, data);               // write new file
+            }
+        }
+        else
+            static assert(0);
+        if (log) printf("same size\n");
+
+        /* The file already exists, and is the same size.
+         * Read it in, and compare for equality.
+         * For larger files, this could be faster by comparing the file
+         * block by block and quitting on first difference.
+         */
+        ReadResult r = read(namez);
+        if (!r.success ||
+            r.buffer.data[] != data[])
+            return write(namez, data); // contents not same, so write new file
```

Comment:
This is not quite optimal.

- The target file will be opened two times, first for reading and then for writing. Ideally it's only open once for reading AND writing. I doubt this makes a measurable difference in practice save for exceptional cases (slow networked drives).
- Reading the entire file before comparing is indeed inefficient, as the comment notes. However, this is only done if the files have the same size so there's a good likelihood they have the same contents.

For a first shot this should be workable.

Reply:
One alternative would be to simply compute a hash of the object file every time it is written and then use that to compare to newer builds (only if the file sizes match). Computing the hash for the new object file should be faster than reading the contents of the old one, however, compared to the block-by-block method proposed by @andralex it will still perform more writes.

The hash can be stored in a file that the compiler will generate for each working directory.